fl4health/clients/tabular_data_client.py

```diff
@@ -0,0 +1,105 @@
+from logging import INFO
+from pathlib import Path
+from typing import Dict, Sequence
+
+import pandas as pd
+import torch
+from flwr.common.logger import log
+from flwr.common.typing import Config, NDArray, Scalar
+
+from fl4health.clients.basic_client import BasicClient
+from fl4health.feature_alignment.constants import FEATURE_INFO, FORMAT_SPECIFIED, INPUT_DIMENSION, OUTPUT_DIMENSION
+from fl4health.feature_alignment.tab_features_info_encoder import TabularFeaturesInfoEncoder
+from fl4health.feature_alignment.tab_features_preprocessor import TabularFeaturesPreprocessor
+from fl4health.utils.metrics import Metric
+
+
+class TabularDataClient(BasicClient):
+    def __init__(
+        self, data_path: Path, metrics: Sequence[Metric], device: torch.device, id_column: str, target_column: str
+    ) -> None:
+        super().__init__(data_path, metrics, device)
+        self.tabular_features_info_encoder: TabularFeaturesInfoEncoder
+        self.tabular_features_preprocessor: TabularFeaturesPreprocessor
+        self.df: pd.DataFrame
+        self.input_dimension: int
+        self.output_dimension: int
+        self.id_column = id_column
+        self.target_column = target_column
+        # The aligned data and targets, which are used to construct dataloaders.
+        self.aligned_features: NDArray
+        self.aligned_targets: NDArray
+
+    def setup_client(self, config: Config) -> None:
+        """
+        Initialize the client by encoding the information of its tabular data
+        and initializing the corresponding TabularFeaturesPreprocessor.
+
+        config[FORMAT_SPECIFIED] indicates whether the server has obtained
+        the source of information to perform feature alignment.
+        If it is True, it means the server has obtained such information
+        (either a priori or by polling a client).
+        So the client will encode that information and use it instead
+        to perform feature preprocessing.
+        """
+        format_specified = self.narrow_config_type(config, FORMAT_SPECIFIED, bool)
+        self.df = self.get_data_frame(config)
+
+        if format_specified:
+            # Since the server has obtained its source of information,
+            # the client will encode that instead.
+            self.tabular_features_info_encoder = TabularFeaturesInfoEncoder.from_json(
+                self.narrow_config_type(config, FEATURE_INFO, str)
+            )
+            self.tabular_features_preprocessor = TabularFeaturesPreprocessor(self.tabular_features_info_encoder)
+
+            # preprocess features.
+            self.aligned_features, self.aligned_targets = self.tabular_features_preprocessor.preprocess_features(
+                self.df
+            )
+
+            # Obtain the input and output dimensions to be sent to
+            # the server for global model initialization.
```

Comment:
We're making an assumption here that the first dimension is the number of rows. Very reasonable, but maybe worth adding to the comment.

Reply:
Super minor but is it worth having a `get_feature_dimension` function in the `tabular_features_info_encoder` class that abstracts this away?

Reply:
I looked into this and it might be a little tricky because the feature dimension is only known after the pipeline is applied to the dataframe... There might be a way to "predict" what the dimension would be based on the type of column transformer applied, but I'm not sure how to do that...

Reply:
Gotcha. That's good with me. Just wanted to see if it were doable.